internal/util/password/scramsha256.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package password
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+
+	"github.com/xdg-go/scram"
+
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+)
+
+// scramSHA256Params represent password parameters for SCRAM-SHA-256 authentication.
+type scramSHA256Params struct {
+	iterationCount int
+	saltLen        int
+}
+
+// fixedScramSHA256Params represent fixed password parameters for SCRAM-SHA-256 authentication.
+var fixedScramSHA256Params = &scramSHA256Params{
+	iterationCount: 15_000,
+	saltLen:        45,
+}
+
+// scramSHA256HashParams hashes the password with the given salt and parameters,
+// and returns the document that should be stored.
+func scramSHA256HashParams(password string, salt []byte, params *scramSHA256Params) (*types.Document, error) {
+	if len(salt) != int(params.saltLen) {
+		return nil, lazyerrors.Errorf("unexpected salt length: %d", len(salt))
+	}
+
+	// username is omitted because it is not used in the hash computation
```

Comment:
This confuses me, client is created only using password? Then its stored credential has no information about its username :thinking:

Reply:
@chilagrow,

Yes, that client wouldn't work to establish a connection. I removed the username from the public API as it's not needed and we discard this client immediately after password generation.

The client is only used there as this library doesn't provide a function for hashing the password directly, without creating one.

Instead of using

```go
// username is omitted because it is not used in the hash computation
client, err := scram.SHA256.NewClient("", password, "")
if err != nil {
	return nil, lazyerrors.Error(err)
}

kf := scram.KeyFactors{
	Salt:  string(salt),
	Iters: params.iterationCount,
}

cred := client.GetStoredCredentials(kf)
```

We could also use something like this, avoiding using this library altogether:

```go
func computeHMAC( key, data []byte) []byte {
	mac := hmac.New(sha256.New, key)
	mac.Write(data)
	return mac.Sum(nil)
}

func computeHash(b []byte) []byte {
	h := sha256.New()
	h.Write(b)
	return h.Sum(nil)
}

var prepPassword string

if prepPassword, err = stringprep.SASLprep.Prepare(password); err != nil {
	return nil, fmt.Errorf("Error SASLprepping password '%s': %v", password, err)
}

saltedPassword := pbkdf2.Key([]byte(prepPassword), []byte(kf.Salt), kf.Iters, 32, aha256.New)
clientKey := computeHMAC(saltedPassword, []byte("Client Key"))
storedKey := computeHash(clientKey)
serverKey := computeHMAC(saltedPassword, []byte("Server Key"))
```

Another option would be to try to open a pull request upstream creating a function for that, but that might take time or not get accepted.

What do you prefer? Either is fine, and I'd usually prefer the second choice, despite being more complex.

Reply:
Also notice that this function only generates the password credentials, and the callee that uses it is where the data is actually recorded.

Reply:
I see, we are using that library to just hash the password. Now your comment makes sense. I agree with you, second choice is preferable, even complex. Because the usage of library misguides creating client to establish connection which we are not doing. :see_no_evil:

Reply:
Going to do it now then.

Reply:
@chilagrow, you can take a look now. I've done the change :)